### DISCUSSION TITLE
Bug: using custom serverless runtime from git or path is not working

### DISCUSSION THREAD

**Discussion by oscartbeaumont** [vercel/vercel#8387] [Help]
On [this page](https://github.com/vercel/vercel/blob/main/DEVELOPING_A_RUNTIME.md) it is stated that

> Official Runtimes are published to the npm registry as a package and referenced in the use property of the vercel.json configuration file.

This however doesn't seem to be the case as if the file path or git url doesn't end in a version the CLI throws the error
> Error! Function Runtimes must have a valid version, for example `now-php@1.0.0`.

My not working `vercel.json` looks like:
```json
{
  "version": 2,
  "public": false,
  "functions": {
    "api/index.py": {
      "runtime": "/Users/oscar/Desktop/vercel/packages/python",
    }
  }
}

```

I tested and if I run `mv my_runtime my_runtime@0.0.1` and then refer to the runtime with an absolute file path which now ends with `my_runtime@0.0.1` it imports the runtime correctly.

The check for a version at the end of the `runtime` field in the `vercel.json` should be skipped if a file path or git url is detected to fix this bug.

**Comment by TooTallNate** [comment]
Hi there @oscartbeaumont, thanks for your question. It's indeed correct that the `runtime` field must point to a specific version of an npm package. Using a file path or git URL will not work.

Can I ask what you are trying to do here? Are you trying to use a modified version of the `@vercel/python` runtime?

**Comment by oscartbeaumont** [comment]
Yes, I was working on fixing [some bugs](https://github.com/vercel/vercel/pull/8384) in the Python runtime and wanted to test my fixes on the project which I was working on when I found the bugs. I couldn't using linking because the Vercel repository was installed with yarn following your contributing guide instructions and my other project uses pnpm. I didn't try installing your codebase with pnpm so I have no idea if that would have worked or not but for a quick one-off test I find file paths a lot easier.

You should update the docs I linked above if there is no intention of classing this as a bug. Deciding to not support git urls and relative file paths when you basically already have support for it seems a bit backward to me. How would someone test a modified version of one of the runtimes in the Vercel production environment if they wanted to? A git url would be perfect for this. Currently, as far as I am aware, they would be required to publish the runtime to npm which makes no sense for a test, especially given pnpm's policies on unpublishing.

**Reply by TooTallNate** [reply]
Ok, so in that case you should be using the version of CLI that is in the `vercel/vercel` repository. That way the changes to the Python runtime will be utilized. I suggest using the `vercel build` command locally to generate build outputs that can then be deployed via `vercel deploy --prebuilt`.

So first remove the "functions" configuration from `vercel.json` - it's not needed for what you're trying to do, and then try these steps:

1. Make your modifications to the Runtime (edit `packages/python/src/index.ts` or whatever)
2. Build the Runtime (run `cd packages/python && yarn build`)
3. Go to the CLI directory (`cd packages/cli`)
4. Run `vercel build` command of the CLI source code (`yarn dev --cwd /path/to/your/test/project build`) - the modified Python runtime will be used
5. Create a deployment from the build artifacts to test on Vercel (`yarn dev --cwd /path/to/your/test/project deploy --prebuilt`)

So no need to mess around with file paths, git URL, nor npm publishing. Let me know if that makes sense 😃
